platforms/android/tangram/src/main/java/com/mapzen/tangram/MapView.java

```diff
@@ -18,45 +19,130 @@
-     * @param listener The listener to receive to receive scene load events;
+     * @param readyCallback {@link MapReadyCallback#onMapReady(MapController)} to be invoked when
+     * {@link MapController} is instantiated and ready to be used. The callback will be made on the UI thread
+     * @param sceneLoadListener The listener to receive to receive scene load events;
      * the callback will be made on the UI thread
+     * @return An instance of {@link MapController} if previously created else, null
      */
-    @NonNull
-    public MapController getMap(@Nullable final MapController.SceneLoadListener listener) {
-        return getMap(listener, null);
+    @Nullable
+    public MapController getMapAsync(@Nullable final MapReadyCallback readyCallback,
+                            @Nullable final MapController.SceneLoadListener sceneLoadListener) {
+        return getMapAsync(readyCallback, sceneLoadListener, null);
     }
 
     /**
-     * Construct a {@code MapController}; may only be called from the UI thread
-     * @param listener The listener to receive to receive scene load events;
+     * Construct a {@code MapController} in an async thread; may only be called from the UI thread
+     * Map instance uses {@link DefaultHttpHandler} for retrieving remote map resources
+     * @param readyCallback {@link MapReadyCallback#onMapReady(MapController)} to be invoked when
+     * {@link MapController} is instantiated and ready to be used. The callback will be made on the UI thread
+     * @param sceneLoadListener The listener to receive to receive scene load events;
      * @param handler Set the client implemented {@link HttpHandler} for retrieving remote map resources
      *                when null {@link DefaultHttpHandler} is used
-     * the callback will be made on the UI thread
+     * @return An instance of {@link MapController} if previously created else, null
      */
-    @NonNull
-    public MapController getMap(@Nullable final MapController.SceneLoadListener listener, @Nullable final HttpHandler handler) {
-        if (mapController != null) {
-            return mapController;
+    @Nullable
+    public MapController getMapAsync(@Nullable final MapReadyCallback readyCallback,
+                            @Nullable final MapController.SceneLoadListener sceneLoadListener,
+                            @Nullable final HttpHandler handler) {
+
+        synchronized (lock) {
+            if (mapController != null) {
+                return mapController;
+            }
+        }
+
+        disposeMapReadyTask();
+        if (glSurfaceView == null) {
+            glSurfaceView = new GLSurfaceView(getContext());
         }
-        mapController = getMapInstance();
-        mapController.setSceneLoadListener(listener);
-        mapController.init(handler);
+        final Context ctx = glSurfaceView.getContext();
+
+        getMapAsyncTask = new AsyncTask<Void, Void, MapController>() {
```

Comment:
Android Studio shows a warning here about creating an anonymous inner class here - it seems like this could lead to leaks. We should probably make an explicit inner class for this instead.

Reply:
Ahh, I see the issue here. 
Reference: https://stackoverflow.com/a/46166223 and https://www.androiddesignpatterns.com/2013/01/inner-class-handler-memory-leak.html